proposed/2021/TargetFrameworkAndPlatformsFilters.md

```diff
@@ -0,0 +1,205 @@
+
+#### Version Numbers
+
+There are many unique version numbers for all generations of .NET & supported platforms. The filtering experience should be smart enough to create a version list based on the current selection. Here are a few scenarios:
+
+- User selects `Any` for the .NET generation.
+  - Version filter populates all known versions for .NET Framework, .NET Core, .NET, and .NET Standard.
+    - 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 2.0, 2.1, 2.2, 3.0, 3.1, 3.5, 4.0, 4.5, 4.5.1, 4.5.2, 4.6, 4.6.1, 4.6.2, 4.7, 4.7.1, 4.7.2, 4.8, 5.0, 6.0
+- User selects `Any` for the .NET generation and `Windows` for the supported platform.
+  - Version filter populates all known version for .NET Framework, .NET Core, .NET, and .NET Standard.
+  - 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 2.0, 2.1, 2.2, 3.0, 3.1, 3.5, 4.0, 4.5, 4.5.1, 4.5.2, 4.6, 4.6.1, 4.6.2, 4.7, 4.7.1, 4.7.2, 4.8, 5.0, 6.0
+- User selects `Any` for the .NET generation and `Android` for the supported platform.
+  - Version filter populates all known version for .NET and .NET Standard.
+    - 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 2.0, 2.1, 6.0
+- User selects `.NET Framework` for the .NET generation.
+  - Version filter populates all known version for .NET Framework and .NET Standard.
+    - 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 2.0, 2.1, 3.5, 4.0, 4.5, 4.5.1, 4.5.2, 4.6, 4.6.1, 4.6.2, 4.7, 4.7.1, 4.7.2, 4.8
+- User selects `.NET Framework` for the .NET generation and `Windows` for the supported platform.
+  - Version filter populates all known version for .NET Framework and .NET Standard.
+    - 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 2.0, 2.1, 3.5, 4.0, 4.5, 4.5.1, 4.5.2, 4.6, 4.6.1, 4.6.2, 4.7, 4.7.1, 4.7.2, 4.8
+- User selects `.NET Core` for the .NET generation and `macOS` for the supported platform.
+  - Version filter populates all known version for .NET Framework and .NET Standard.
+    - 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 2.0, 2.1, 2.2, 3.0, 3.1
+
+For further scenarios, see the [framework precedence](https://nugettools.azurewebsites.net/6.0.0-preview.3/framework-precedence) tool to understand how the version matrix would be calculated.
+
+![](meta/resources/TargetFrameworkAndPlatformFilters/VersionNumbers.png)
+
+**Note:** The label should best represent the ".NET" version. Thus the label name may change from "versions" to "framework versions / .NET versions" or similar.
+
+#### Complete Experience
+
+![](meta/resources/TargetFrameworkAndPlatformFilters/CompleteExperience.png)
+
+**Note:** There's opportunity to put a tooltip in the complete experience to help explain what the difference of .NET, .NET Core, .NET Framework is & link to a documentaiton page.
+
+## Search Syntax
+
+Lastly, NuGet should add a protocol in which supports [search syntax](https://docs.microsoft.com/nuget/consume-packages/finding-and-choosing-packages#search-syntax) for frameworks, platforms, and potentially *implementations**.
+
+Here's an ideal `<property>:<term>` for each use-case:
```

Comment:
I am concerned about making the filtering syntax based on the `q=` (query term) parameter to search instead of a different query parameter (similar to what @zivkan is alluding to in https://github.com/NuGet/Home/pull/11374#discussion_r748346221). This search term will not be recognized by other search implementations (e.g. Azure DevOps) and may lead to bad results on those feeds. This is mainly relevant with the "All" source selected in VS PM UI, browse tab.

If we make this filter a query string (like `framework=X&q=json`) it will be ignored by other search services. Ignoring unknown query parameters is very standard in REST/HTTP whereas the parsing of `q=` (the search term) is much less forgiving or standardized.

Take for example, this URL:
https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/9d15d80a-6afc-4f7e-901b-9378146a4b8b/nuget/v3/query2/?q=foo%20framework:net6.0
This is search "foo framework:net6.0" on an a public Azure DevOps feed. **Zero results.**

Compare this to:
https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/9d15d80a-6afc-4f7e-901b-9378146a4b8b/nuget/v3/query2/?q=foo&0framework=net6.0
This is sending a `framework=net6.0` query parameter to the same feed. **One result.**

There is a whole technical discussion needed on what to do with feeds of mixed capabilities (nuget.org having framework filtering and Azure DevOps not yet having it) but I fear that this approach of hijacking the search text for this will limit our solution space or force us into nasty hacks.

We _could_ say that this syntax only works on NuGet.org UI but this would be a change in patterns. Today, any search you can do on nuget.org can be done similarly in PM UI to get the same results.

As an aside, this approach also diverges from the implementation of other nuget.org advanced filters, e.g. sorting and package type filtering. These are both handled with query parameters rather than adding `<property>:<term>` to the search text.

I see a direction here that leads to an advanced search query building UX on NuGet.org (similar to GitHub's or Google's old page) that builds up the search text with `<property>:<term>` specifiers, and I like that idea and direction, but it's hard for me to see how this would work well in the near tear where there is little consistency in search syntax across feeds.

Reply:
I don't have strong opinions because I'm quite ignorant about search. I'm hoping we can first define the frontend experience for the user via search syntax that they can use alternative to clicking through filters & we do the right thing in the backend to generate the proper search result.

I used the current experience of typing in something like `authors:microsoft` into NuGet.org as a reference & that in turn would generate the respective query(q=?) parameters. If it's more ideal to follow the advanced filters paradigm of `framework=` then by all means! 

The goal at the end of the day would be to allow someone to apply filters, possibly generate those in the search box itself as hint text, and give the user the respective results with those filters applied.

Reply:
If the user experience requires typing into the search box, we may need to do weird hacks like parsing out framework/platform/implementation search tokens and converting them behind the scenes to query parameters -- at least on the VS side as to not mess with other feeds that don't yet support compatibility filtering. Hopefully we don't have to do that.

I think it's worth adding a note to this section or change emphasis such that the filter experience is the priority and the search text is optional pending feasibility investigation. That's my opinion at least 😄.

On the VS side, I am imagining this filtering is done implicitly based on project context or with a similar filtering UI that's on NuGet.org. The reason I bring this up is that we've attempted to keep reasonable parity between the two search experiences (nuget.org + VS PM UI).

Reply:
> I think it's worth adding a note to this section or change emphasis such that the filter experience is the priority and the search text is optional pending feasibility investigation. That's my opinion at least 😄.

Yep let's do that.

> On the VS side, I am imagining this filtering is done implicitly based on project context or with a similar filtering UI that's on NuGet.org. The reason I bring this up is that we've attempted to keep reasonable parity between the two search experiences (nuget.org + VS PM UI).

Yep, I think right now even with `supportedFramework`, I don't think we can capture a similar browse/search experience in the client until we do the nuget.org side. Even if we matched a list of project frameworks, we may want users to look for platform specific vs. portable assets which isn't possible today I don't believe.